File: 1-working-with-list/p15_replicate.ml

let replicate list rep =
  if rep <= 0 then
    []
  else
    let rec aux acc curr = function
      | [] -> acc
      | (h :: t) as l ->
        if curr == 1 then
          aux (h :: acc) rep t
        else
          aux (h :: acc) (curr-1) l
    in List.rev (aux [] rep list);;

replicate ["a";"b";"c"] 3;;
replicate ["a";"b";"c"] 0;;
replicate ["a";"b";"c"] 1;;
replicate ["a"] 5;;
replicate [] 5;;

File: 1-working-with-list/p15_replicate.scala
import scala.annotation.tailrec

object Main {

  def main(args: Array[String]) {
    val l = List("a", "b", "c")
    println(replicate(l, 3))
  }

  def rep[T](e: T, count: Int): List[T] = {
    @tailrec
    def aux(cur: Int, acc: List[T]): List[T] = {
      if(cur > 0)
        aux(cur-1, e :: acc)
      else
        acc
    }
    aux(count, List())
  }

  def replicate[T](list: List[T], count: Int): List[T] = {
    @tailrec
    def aux(l: List[T], acc: List[T]): List[T] = {
      l match {
        case Nil => acc
        case head :: tail => aux(tail, rep(head, count) ::: acc)
      }
    }
    aux(list, List())
  }
}
